logstash/parallel_process.go

```diff
@@ -0,0 +1,284 @@
+	}
+	ts.senderListener.SetUnlinkOnClose(false)
+
+	go func() {
+		defer close(ts.senderReady)
+
+		ts.sender, err = ts.senderListener.AcceptUnix()
+		if err != nil {
+			log.Errorf("Error while accept unix socket: %s", err)
+		}
+		ts.senderListener.Close()
+	}()
+
+	// Unfortunately Logstash doesn't make it easy to just read
+	// events from a stdout-connected pipe and the log from a
+	// stderr-connected pipe. Stdout can contain other garbage (at
+	// the very least "future logs will be sent to ...") and error
+	// messages could very well be sent there too. Mitigate by
+	// having Logstash write output logs to a temporary file and
+	// its own logs to a different temporary file.
+	outputFile, err := newDeletedTempFile("", "")
+	if err != nil {
+		return nil, err
+	}
+	ts.receiver = outputFile
+
+	return ts, nil
+}
+
+// Write writes to the sender of the TestStream
+func (ts *TestStream) Write(p []byte) (n int, err error) {
+	timer := time.NewTimer(ts.timeout)
+	select {
+	case <-ts.senderReady:
+	case <-timer.C:
+		return 0, fmt.Errorf("Write timeout error")
+	}
+	return ts.sender.Write(p)
+}
+
+// Close closes the sender of the TestStream
+func (ts *TestStream) Close() error {
+	if ts.sender != nil {
+		err := ts.sender.Close()
+		ts.sender = nil
+		return err
+	}
+	return nil
+}
+
+// Cleanup closes and removes all temporary resources
+// for a TestStream
+func (ts *TestStream) Cleanup() {
+	if ts.senderListener != nil {
+		ts.senderListener.Close()
+	}
+	if ts.sender != nil {
+		ts.Close()
+	}
+	os.RemoveAll(ts.senderPath)
+	if ts.receiver != nil {
+		ts.receiver.Close()
+	}
+}
+
+// CleanupTestStreams closes all sockets and streams as well
+// removes temporary file ressources for an array of
+// TestStreams
+func CleanupTestStreams(ts []*TestStream) {
+	for i := range ts {
+		ts[i].Cleanup()
+	}
+}
+
+// ParallelProcess represents the invocation and execution of a Logstash child
+// process that emits JSON events from multiple inputs through filter to multiple outputs
+// configuration files supplied by the caller.
+type ParallelProcess struct {
+	streams []*TestStream
+
+	child     *exec.Cmd
+	configDir *string
+	log       io.ReadCloser
+
+	stdio io.Reader
+}
+
+// NewParallelProcess prepares for the execution of a new Logstash process but
+// doesn't actually start it. logstashPath is the path to the Logstash
+// executable (typically /opt/logstash/bin/logstash). The configs parameter is
+// one or more configuration files containing Logstash filters.
+func NewParallelProcess(logstashPath string, testStream []*TestStream, keptEnvVars []string, configs ...string) (*ParallelProcess, error) {
+	if len(configs) == 0 {
+		return nil, errors.New("must provide non-empty list of configuration file or directory names")
+	}
+
+	logstashInput := make([]string, len(testStream))
+	logstashOutput := make([]string, len(testStream))
+
+	for i, sp := range testStream {
+		sp.fields["@metadata"] = map[string]interface{}{"testcase": strconv.Itoa(i)}
```

Comment:
Could we make the name of the subfield a bit more unique to reduce the risk of collisions with metadata fields used by the filters under test? Like "__lfv_testcase" or similar.